src/hotspot/share/gc/serial/serialFullGC.cpp

```diff
@@ -266,8 +266,7 @@ class Compacter {
     _spaces[1].init(heap->young_gen()->eden());
     _spaces[2].init(heap->young_gen()->from());
 
-    bool is_promotion_failed = (heap->young_gen()->from()->next_compaction_space() != nullptr);
-    if (is_promotion_failed) {
+    if (!heap->young_gen()->to()->is_empty()) {
```

Comment:
A local variable `bool promotion_failed = ...` and using it afterwards would improve readability.

Reply:
Actually, it is not only about failed promotion. When the to-space is not empty, we must deal with it. But usually, the failed promotion causes that the to-space is not empty.

Reply:
IMO, I think it is good to say: if the to-space is not empty, we need to deal with it, regardless whether it is failed promotion. Even only the failed promotion causes non empty to-space.

What do you think about it?

Reply:
I have no preference regarding using a local-var or not.

(My understanding is that non-empty-to-space == promotion-failure; if that's not true, I wanna refresh my knowledge on that.)

Reply:
That is true, but since the only way this can happen is during promotion failure, the additional local var acts as a documentation about the expected situation this can happen. It isn't like that that `heap->young_gen()->to()->is_empty()` is lost because of that.
Also it's a statement with a triple indirection, so it would be nice to factor it out, whatever the name is, e.g.
```
  bool to_space_empty = heap->young_gen()->to()->is_empty();
  if (to_space_empty) {
```
or as suggested
```
  bool promotion_failed = heap->young_gen()->to()->is_empty();
  if (promotion_failed) {
```
seems better to me. I'll let the decision be up to you.